File: src/interval/IntervalDiatonic.ts
import { ImmutablesCache } from '../common/ImmutablesCache';
import { Diatonic } from '../degrees/Diatonic';
import { IntervalSymbolic } from './Interval';

export class IntervalDiatonic implements IntervalSymbolic {
    public static UNISON;
    public static SECOND;
    public static THIRD;
    public static FOURTH;
    public static FIFTH;
    public static SIXTH;
    public static SEVENTH;
    public static OCTAVE;
    public static NINTH;
    public static TENTH;
    public static ELEVENTH;
    public static TWELFTH;
    public static THIRTEENTH;
    public static FOURTEENTH;
    public static FIFTEENTH;

    private static immutablesCache = new ImmutablesCache<IntervalDiatonic, number>(
        function (num: number): string {
            return "" + num;
        },
        function (intervalDiatonic: IntervalDiatonic): number {
            return intervalDiatonic.number;
        },
        function (num: number): IntervalDiatonic {
            return new IntervalDiatonic(num);
        }
    );

    private constructor(private _number: number) {
    }

    public static from(num: number) {
        if (num < 0)
            num = num % Diatonic.NUMBER + Diatonic.NUMBER;
        return IntervalDiatonic.immutablesCache.getOrCreate(num);
    }

    get number(): number {
        return this._number;
    }

    public toString(): string {
        switch (this) {
            case IntervalDiatonic.UNISON: return "UNISON";
            case IntervalDiatonic.SECOND: return "SECOND";
            case IntervalDiatonic.THIRD: return "THIRD";
            case IntervalDiatonic.FOURTH: return "FOURTH";
            case IntervalDiatonic.FIFTH: return "FIFTH";
            case IntervalDiatonic.SIXTH: return "SIXTH";
            case IntervalDiatonic.SEVENTH: return "SEVENTH";
            case IntervalDiatonic.OCTAVE: return "OCTAVE";
            case IntervalDiatonic.NINTH: return "NINTH";
            case IntervalDiatonic.TENTH: return "TENTH";
            case IntervalDiatonic.ELEVENTH: return "ELEVENTH";
            case IntervalDiatonic.TWELFTH: return "TWELFTH";
            case IntervalDiatonic.THIRTEENTH: return "THIRTEENTH";
            case IntervalDiatonic.FOURTEENTH: return "FOURTEENTH";
            case IntervalDiatonic.FIFTEENTH: return "FIFTEENTH";
        }

        return "IntervalDiatonic: " + this.number;
    }

    private static initialize() {
        this.UNISON = IntervalDiatonic.from(0);
        this.SECOND = IntervalDiatonic.from(1);
        this.THIRD = IntervalDiatonic.from(2);
        this.FOURTH = IntervalDiatonic.from(3);
        this.FIFTH = IntervalDiatonic.from(4);
        this.SIXTH = IntervalDiatonic.from(5);
        this.SEVENTH = IntervalDiatonic.from(6);
        this.OCTAVE = IntervalDiatonic.from(7);
        this.NINTH = IntervalDiatonic.from(8);
        this.TENTH = IntervalDiatonic.from(9);
        this.ELEVENTH = IntervalDiatonic.from(10);
        this.TWELFTH = IntervalDiatonic.from(11);
        this.THIRTEENTH = IntervalDiatonic.from(12);
        this.FOURTEENTH = IntervalDiatonic.from(13);
        this.FIFTEENTH = IntervalDiatonic.from(14);
    }
}

File: src/main/java/es/danisales/datune/interval/IntervalDiatonic.java
package es.danisales.datune.interval;

import es.danisales.datune.degrees.octave.Diatonic;
import es.danisales.datune.degrees.scale.DiatonicDegree;
import es.danisales.datune.lang.Nominator;
import es.danisales.utils.NeverHappensException;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

public enum IntervalDiatonic implements Interval {
    UNISON, SECOND, THIRD, FOURTH, FIFTH, SIXTH, SEVENTH,
    OCTAVE, NINTH, TENTH, ELEVENTH, TWELFTH, THIRTEENTH, FOURTEENTH,
    FIFTEENTH;

    public static @NonNull IntervalDiatonic fromIndex(int n) {
    	n = trimToValidValue(n);

    	return IntervalDiatonic.values()[n];
    }

    private static int trimToValidValue(int n) {
        return (n < 0 ? -n : n) % (IntervalDiatonic.values().length);
    }

    @SuppressWarnings("Duplicates")
    public static @NonNull IntervalDiatonic from(@NonNull DiatonicDegree diatonicDegree) {
        switch (diatonicDegree) {
            case I:
                return UNISON;
            case II:
                return SECOND;
            case III:
                return THIRD;
            case IV:
                return FOURTH;
            case V:
                return FIFTH;
            case VI:
                return SIXTH;
            case VII:
                return SEVENTH;
        }

        throw NeverHappensException.switchOf(diatonicDegree);
    }

    public static @NonNull IntervalDiatonic from(@NonNull IntervalChromatic intervalChromatic) {
        switch (intervalChromatic) {
            case AUGMENTED_UNISON:
            case PERFECT_UNISON:
                return IntervalDiatonic.UNISON;
            case AUGMENTED_SECOND:
            case DIMINISHED_SECOND:
            case MAJOR_SECOND:
            case MINOR_SECOND:
                return IntervalDiatonic.SECOND;
            case AUGMENTED_THIRD:
            case DIMINISHED_THIRD:
            case MAJOR_THIRD:
            case MINOR_THIRD:
                return IntervalDiatonic.THIRD;
            case AUGMENTED_FOURTH:
            case DIMINISHED_FOURTH:
            case PERFECT_FOURTH:
                return IntervalDiatonic.FOURTH;
            case AUGMENTED_FIFTH:
            case DIMINISHED_FIFTH:
            case PERFECT_FIFTH:
                return IntervalDiatonic.FIFTH;
            case AUGMENTED_SIXTH:
            case DIMINISHED_SIXTH:
            case MAJOR_SIXTH:
            case MINOR_SIXTH:
                return IntervalDiatonic.SIXTH;
            case AUGMENTED_SEVENTH:
            case DIMINISHED_SEVENTH:
            case MAJOR_SEVENTH:
            case MINOR_SEVENTH:
                return IntervalDiatonic.SEVENTH;
            case DIMINISHED_OCTAVE:
            case PERFECT_OCTAVE:
            case AUGMENTED_OCTAVE:
                return IntervalDiatonic.OCTAVE;
            case AUGMENTED_NINTH:
            case DIMINISHED_NINTH:
            case MAJOR_NINTH:
            case MINOR_NINTH:
                return IntervalDiatonic.NINTH;
            case AUGMENTED_TENTH:
            case DIMINISHED_TENTH:
            case MAJOR_TENTH:
            case MINOR_TENTH:
                return IntervalDiatonic.TENTH;
            case AUGMENTED_ELEVENTH:
            case DIMINISHED_ELEVENTH:
            case PERFECT_ELEVENTH:
                return IntervalDiatonic.ELEVENTH;
            case AUGMENTED_TWELFTH:
            case DIMINISHED_TWELFTH:
            case PERFECT_TWELFTH:
                return IntervalDiatonic.TWELFTH;
            case AUGMENTED_THIRTEENTH:
            case DIMINISHED_THIRTEENTH:
            case MAJOR_THIRTEENTH:
            case MINOR_THIRTEENTH:
                return IntervalDiatonic.THIRTEENTH;
            case AUGMENTED_FOURTEENTH:
            case DIMINISHED_FOURTEENTH:
            case MAJOR_FOURTEENTH:
            case MINOR_FOURTEENTH:
                return IntervalDiatonic.FOURTEENTH;
            case DIMINISHED_FIFTEENTH:
            case PERFECT_FIFTEENTH:
            case AUGMENTED_FIFTEENTH:
                return IntervalDiatonic.FIFTEENTH;
        }

        throw NeverHappensException.switchOf(intervalChromatic);
    }

    public String toString() {
        return "Intervalo de " + Nominator.from(this);
    }

    @Override
    public boolean isCompound() {
        return ordinal() > Diatonic.NUMBER;
    }

    public @Nullable IntervalDiatonic getNext() {
        int index = ordinal() + 1;
        if (index >= values().length)
            return null;
        return values()[index];
    }

    public @Nullable IntervalDiatonic getPrevious() {
        int index = ordinal() - 1;
        if (index < 0)
            return null;
        return values()[index];
    }
}
